File: app/controllers/submitted_proposals_controller.rb
class SubmittedProposalsController < ApplicationController
  before_action :authenticate_user!
  before_action :authorize_user
  before_action :set_proposals, only: %i[index]
  before_action :set_proposal, except: %i[index download_csv import_reviews
                                          reviews_booklet reviews_excel_booklet]
  before_action :template_params, only: %i[approve_decline_proposals]
  before_action :check_reviews_permissions, only: %i[import_reviews
                                                     reviews_booklet
                                                     reviews_excel_booklet]

  def index; end

  def show
    @proposal.review! if @proposal.may_review?
    log_activity(@proposal)
  end

  def edit
    @proposal.invites.build
  end

  def download_csv
    @proposals = Proposal.where(id: params[:ids].split(','))
    log_activities(@proposals)
    send_data Proposal.to_csv(@proposals), filename: "submitted_proposals.csv"
  end

  def edit_flow
    params[:ids]&.split(',')&.each do |id|
      @proposal = Proposal.find_by(id: id.to_i)
      check_proposal_status and return unless @proposal.may_progress?

      break unless post_to_editflow
    end

    respond_to do |format|
      format.js { render js: "window.location='/submitted_proposals'" }
      message = "Proposals submitted to EditFlow!"
      format.html { redirect_to submitted_proposals_path, notice: message }
    end
  end

  def revise_proposal_editflow
    @proposal = Proposal.find_by(id: params[:proposal_id].to_i)
    unless @proposal.may_progress_spc?
      redirect_to versions_proposal_url(@proposal),
                  alert: "Proposal status should be initial_review or revision_submitted_spc."
      return
    end
    check_proposal_editflow_id
    nil
  end

  def staff_discussion
    @staff_discussion = StaffDiscussion.new
    discussion = params[:discussion]
    if @staff_discussion.update(discussion: discussion,
                                proposal_id: @proposal.id)
      redirect_to submitted_proposal_url(@proposal),
                  notice: "Your comment was added!"
    else
      redirect_to submitted_proposal_url(@proposal),
                  alert: @staff_discussion.errors.full_messages
    end
  end

  def send_emails
    raise CanCan::AccessDenied unless @ability.can?(:manage, Email)

    @email = Email.new(email_params.merge(proposal_id: @proposal.id))
    change_status
    unless @check_status
      @message = "Proposal status cannot be changed!"
      page_redirect_with_alert
      return
    end
    add_files
    organizers_email_addresses
    if @email.save
      @email.email_organizers(@organizers_email)
      page_redirect
    else
      @message = @email.errors.full_messages
      page_redirect_with_alert
    end
  end

  def destroy
    @proposal.destroy
    respond_to do |format|
      format.html do
        redirect_to submitted_proposals_url,
                    notice: "Proposal was successfully deleted."
      end
      format.json { head :no_content }
    end
  end

  def approve_decline_proposals
    params[:proposal_ids]&.split(',')&.each do |id|
      create_birs_email(id)
      @errors = []
      render_error and return unless @check_status

      send_email_proposals
    end
    if @errors.empty?
      head :ok
    else
      render json: @errors.flatten.to_json, status: :unprocessable_entity
    end
  end

  def proposals_booklet
    proposal_ids = params[:proposal_ids]
    table = params[:table]
    counter = proposal_ids.split(',').count
    ProposalBookletJob.perform_later(proposal_ids, table, counter, current_user)
    head :accepted
  end

  def download_booklet
    f = File.open(Rails.root.join('tmp/booklet-proposals.pdf'))
    send_file(
      f,
      filename: "proposal_booklet.pdf",
      type: "application/pdf"
    )
  end

  def table_of_content
    proposals = params[:proposals]
    render json: { proposals: proposals }, status: :ok
  end

  def update_status
    status = params[:status]
    if @proposal.update(status: status.to_i)
      render json: {}, status: :ok
    else
      render json: @proposal.errors.full_messages, status: :unprocessable_entity
    end
  end

  def update_location
    location = params[:location]
    if @proposal.update(assigned_location_id: location)
      render json: {}, status: :ok
    else
      render json: @proposal.errors.full_messages, location: :unprocessable_entity
    end
  end

  def import_reviews
    ImportJob.perform_later(params[:proposals], current_user, params[:action])
    head :accepted
  end

  def reviews_booklet
    temp_file = "propfile-#{current_user.id}-review-booklet.tex"
    ReviewJob.perform_now(params[:proposals], params[:reviewContentType],
                          params[:table], current_user, temp_file)
    head :accepted
  end

  def download_review_booklet
    pdf_file = Rails.root.join("tmp/proposal-reviews-#{current_user.id}.pdf")
    year = Date.current.year + 2
    filename = "#{year}-proposal-reviews.pdf"
    if File.exist?(pdf_file)
      file = File.open(pdf_file)
      send_file(
        file,
        filename: filename,
        type: "application/pdf"
      )
    else
      render json: "File not found: #{pdf_file}"
    end
  end

  def reviews; end

  def reviews_excel_booklet
    check_selected_proposals
    @proposals = Proposal.where(id: params[:proposals].split(','))
    log_activities(@proposals)
    respond_to do |format|
      format.xlsx
    end
  end

  def proposal_outcome_location
    proposals = Proposal.where(selected_proposal_ids)
    proposals.update_all(outcome_location_params.to_hash)
    head :ok
  end

  private

  def selected_proposal_ids
    params.require(:proposal).permit(id: [])
  end

  # def query_params?
  #   params.values.any?(&:present?)
  # end

  def outcome_location_params
    params.require(:proposal).permit(:outcome, :assigned_location_id, :assigned_size)
  end

  def email_params
    params.permit(:subject, :body, :cc_email, :bcc_email)
  end

  def set_proposals
    @proposals = Proposal.order(:code, :created_at)
    @proposals = ProposalFiltersQuery.new(@proposals).find(params)
  end

  def change_status
    revision_template
    @check_status = @email.update_status(@proposal, 'Approval') if params[:templates].split(':').first == "Approval"
    @check_status = @email.update_status(@proposal, 'Reject') if params[:templates].split(':').first == "Reject"
    @check_status = @email.update_status(@proposal, 'Decision') if params[:templates].split.first == "Decision"
  end

  def revision_template
    case params[:templates].split(':').first
    when "Revision"
      @check_status = @email.update_status(@proposal, 'Revision')
    when "Revision SPC"
      @check_status = @email.update_status(@proposal, 'Revision SPC')
    end
  end

  def latex_temp_file
    "propfile-#{current_user.id}-#{@proposal.id}.tex"
  end

  def generate_pdf_string
    pdf_file = render_to_string layout: "application", inline: @prop_latex, formats: [:pdf]
    write_pdf_file(pdf_file)
  rescue StandardError => e
    Rails.logger.info { "\n\n#{@proposal.code} LaTeX error:\n #{e.message}\n\n" }
    flash[:alert] = "#{@proposal.code} LaTeX error: #{e.message}"
    @errors = "#{@proposal.code} LaTeX error: #{e.message}"
    ''
  end

  def write_pdf_file(pdf_file)
    File.open(@pdf_path, "w:UTF-8") do |file|
      file.write(pdf_file)
    end
  rescue StandardError => e
    Rails.logger.info { "\n\nError creating #{@proposal&.code} PDF: #{e.message}\n\n" }
    flash[:alert] = "Error creating #{@proposal&.code} PDF: #{e.message}"
    false
  end

  def post_to_editflow
    log_info
    return unless create_pdf_file

    begin
      edit_flow_query = EditFlowService.new(@proposal).query
    rescue RuntimeError => e
      Rails.logger.info { "\n\nErrors in #{@proposal.code}: #{e.message}\n\n" }
      flash[:alert] = "Errors in #{@proposal.code}: #{e.message}"
    end
    return if flash[:alert].present?

    response = RestClient.post ENV.fetch('EDITFLOW_API_URL', nil),
                               { query: edit_flow_query, fileMain: File.open(@pdf_path) },
                               { x_editflow_api_token: ENV.fetch('EDITFLOW_API_TOKEN', nil) }

    query_response_body(response)
    Rails.logger.info { "\n\n*****************************************\n\n" }
    true
  end

  def create_pdf_file
    @prop_latex = ProposalPdfService.new(@proposal.id, latex_temp_file, 'all', current_user)
                                    .generate_latex_file.to_s

    @year = @proposal&.year || (Date.current.year.to_i + 2)
    @pdf_path = Rails.root.join('tmp', "#{@proposal&.code}-#{DateTime.now.to_i}.pdf")

    generate_pdf_string
  end

  def query_response_body(response)
    if response.body.include?("errors")
      Rails.logger.info { "\n\nError sending #{@proposal.code}: #{response.body}\n\n" }
      flash[:alert] = "Error sending #{@proposal.code}: #{response.body}"
      nil
    else
      Rails.logger.info { "\n\nEditFlow response: #{response.inspect}\n\n" }
      flash[:notice] = "#{@proposal&.code} sent to EditFlow!"
      store_response_id(response)
      @proposal.progress!
    end
  end

  def store_response_id(response)
    response_body = JSON.parse(response.body)
    article = response_body["data"]["article"]
    @id = article["id"]
    @proposal.update(editflow_id: @id, edit_flow: DateTime.current)
    @proposal_version = @proposal.proposal_versions.find_by(version: 1)
    @proposal_version.update(editflow_id: @id, send_to_editflow: DateTime.current)
  end

  def check_proposal_editflow_id
    if @proposal.editflow_id.blank?
      redirect_to versions_proposal_url(@proposal), alert: "Proposal has not editflow_id!"
    else
      revision_proposal
    end
    nil
  end

  def revision_proposal
    revise_post_to_editflow
    if @errors.present?
      redirect_to versions_proposal_url(@proposal), alert: @errors
    else
      redirect_to versions_proposal_url(@proposal), notice: "Proposal is successfully sent to EditFlow!"
    end
    nil
  end

  def revise_post_to_editflow
    log_info
    @errors = ""
    revise_query = create_file_revise_query

    return if @errors.present?

    return unless cover_letter_file

    response = RestClient.post ENV.fetch('EDITFLOW_API_URL', nil),
                               { query: revise_query, fileMain: File.open(@pdf_path),
                                 fileRevisionLetter: File.open(@letter_path) },
                               { x_editflow_api_token: ENV.fetch('EDITFLOW_API_TOKEN', nil) }

    mutation_response_body(response, params[:version].to_i)
  end

  def create_file_revise_query
    version = params[:version].to_i
    return unless create_revise_proposal_file(version)

    begin
      revise_query = EditFlowService.new(@proposal).revise_article
    rescue RuntimeError => e
      Rails.logger.info { "\n\nErrors in #{@proposal.code}: #{e.message}\n\n" }
      @errors = "Errors in #{@proposal.code}: #{e.message}"
    end
    revise_query
  end

  def create_revise_proposal_file(version)
    @prop_latex = ProposalPdfService.new(@proposal.id, latex_temp_file, 'all', current_user, version)
                                    .generate_latex_file.to_s

    # @year = @proposal&.year || (Date.current.year.to_i + 2)
    @pdf_path = Rails.root.join('tmp', "#{@proposal&.code}-#{DateTime.now.to_i}.pdf")
    generate_pdf_string
  end

  def log_info
    Rails.logger.info { "\n\n*****************************************\n\n" }
    Rails.logger.info { "\n\nPosting #{@proposal.code} to EditFlow...\n\n" }
    Rails.logger.info { "\n\nCreating PDF for #{@proposal&.code}...\n\n" }
  end

  def mutation_response_body(response, version)
    if response.body.include?("errors")
      Rails.logger.info { "\n\nError sending #{@proposal.code}: #{response.body}\n\n" }
      @errors = "Error sending #{@proposal.code}: #{response.body}"
      return
    else
      Rails.logger.info { "\n\nEditFlow response: #{response.inspect}\n\n" }
      store_revised_response_id(response, version)
      @proposal.progress_spc!
    end
    Rails.logger.info { "\n\n*****************************************\n\n" }
  end

  def cover_letter_file
    cover_letter_pdf = cover_letter_field_pdf
    @letter_path = Rails.root.join('tmp', "#{@proposal&.code}-cover_letter.pdf")
    File.open(@letter_path, "w:UTF-8") do |file|
      file.write(cover_letter_pdf)
    end
  rescue StandardError => e
    Rails.logger.info { "\n\nError creating #{@proposal&.code} PDF: #{e.message}\n\n" }
    @errors = "Error creating #{@proposal&.code} PDF: #{e.message}"
    false
  end

  def cover_letter_field_pdf
    latex = if @proposal.cover_letter
              "\\begin{document}\n#{LatexToPdf.escape_latex(@proposal.cover_letter)}"
            else
              "\\begin{document}\n"
            end
    render_to_string layout: "application", inline: latex, formats: [:pdf]
  end

  def store_revised_response_id(response, version)
    response_body = JSON.parse(response.body)
    article = response_body["data"]["articleReviewVersion"]
    id = article["id"]
    proposal_version = @proposal.proposal_versions.find_by(version: version)
    proposal_version&.update(editflow_id: id, send_to_editflow: DateTime.current)
  end

  def set_proposal
    @proposal = Proposal.find_by(id: params[:id])
  end

  def template_params
    templates = params[:templates].split(": ")
    if templates.first == "Decision Email"
      type = "#{templates.first.downcase.tr!(' ', '_')}_type"
    elsif templates.first.present?
      type = "#{templates.first.downcase}_type"
    end
    template = templates.last
    @email_template = EmailTemplate.find_by(email_type: type, title: template)
    @email_template.update(subject: params[:subject], body: params[:body]) if @email_template.present?
  end

  def send_email_proposals
    add_attachments
    organizers_email = @proposal.invites.where(invited_as: 'Organizer', status: :confirmed)&.pluck(:email)
    @email.new_email_organizers(organizers_email) if @email.save
    @errors << @email.errors.full_messages unless @email.errors.empty?
  end

  def authorize_user
    authorize! params[:action], SubmittedProposalsController
  end

  def check_proposal_status
    render json: { errors: 'Please select initial review proposal(s).' }, status: :unprocessable_entity
  end

  def create_birs_email(id)
    @proposal = Proposal.find_by(id: id)
    @email = Email.new(email_params.merge(proposal_id: @proposal.id))
    change_status
  end

  def render_error
    @errors << "Proposal status cannot be changed"
    render json: @errors.flatten.to_json, status: :unprocessable_entity
  end

  def add_files
    params[:files]&.each do |file|
      @email.files.attach(file)
    end
  end

  def add_attachments
    params[:attachments]&.each do |file|
      @email.files.attach(file)
    end
  end

  def page_redirect
    if params[:action] == "show"
      redirect_to submitted_proposal_url(@proposal),
                  notice: t('submit_proposals.page_redirect.alert')
    else
      redirect_to edit_submitted_proposal_url(@proposal),
                  notice: t('submit_proposals.page_redirect.alert')
    end
  end

  def page_redirect_with_alert
    if params[:action] == "show"
      redirect_to submitted_proposal_url(@proposal),
                  alert: @message
    else
      redirect_to edit_submitted_proposal_url(@proposal),
                  alert: @message
    end
  end

  def organizers_email_addresses
    return if params[:organizers_email].blank?

    @organizers_email = JSON.parse(params[:organizers_email]).map(&:values).flatten
  end

  def check_selected_proposals
    @proposal_ids = params[:proposals]
    @no_review_proposal_ids = []
    @review_proposal_ids = []
    check_proposals_reviews
  end

  def check_proposals_reviews
    return if @proposal_ids.blank?

    pids = @proposal_ids.is_a?(String) ? @proposal_ids.split(',') : @proposal_ids
    pids.each do |id|
      @proposal = Proposal.find_by(id: id)
      reviews_conditions
    end
  end

  def reviews_conditions
    if @proposal.reviews.present?
      @review_proposal_ids << @proposal.id
    elsif @proposal.editflow_id.present?
      ImportReviewsService.new(@proposal).proposal_reviews
      @review_proposal_ids << @proposal.id
    else
      @no_review_proposal_ids << @proposal.id
    end
  end

  def log_activities(proposals)
    proposals.map { |proposal| log_activity(proposal) }
  end

  def log_activity(proposal)
    data = {
      logable: proposal,
      user: current_user,
      data: {
        action: params[:action].humanize
      }
    }
    Log.create!(data)
  end

  def check_reviews_permissions
    raise CanCan::AccessDenied unless @ability.can?(:manage, Review)
  end
end

File: app/javascript/controllers/submitted_proposals_controller.js
import { Controller } from "stimulus"
import Rails from '@rails/ujs'
import toastr from 'toastr'
import Tagify from '@yaireo/tagify'

export default class extends Controller {
  static targets = [ 'toc', 'ntoc', 'templates', 'status', 'statusOptions', 'proposalStatus',
                     'organizersEmail', 'bothReviews', 'scientificReviews', 'ediReviews',
                     'reviewToc', 'reviewNToc' ]

  connect () {
    let proposalId = 0
    if (this.hasOrganizersEmailTarget) {
      var inputElm = this.organizersEmailTarget,
      tagify = new Tagify (inputElm);
    }
  }

  editFlow() {
    var proposalIds = [];
    $("input:checked").each(function() {
      proposalIds.push(this.dataset.value);
    });
    if(typeof proposalIds[0] === "undefined")
    {
      toastr.error("Please select any checkbox!")
    }
    else {
      let data = new FormData()
      data.append("ids", proposalIds)
      var url = `/submitted_proposals/edit_flow`
      Rails.ajax({
        url,
        type: "POST",
        data,
        error: (response) => {
          let errors = response.errors
          toastr.error(errors)
        }
      })
    }
  }

  emailTemplate() {
    let value = event.currentTarget.value
    if(value) {
      let data = new FormData()
      data.append("email_template", value)
      var url = `/emails/email_template`
      Rails.ajax({
        url,
        type: "PATCH",
        data,
        success: (data) => {
          $('#birs_email_subject').val(data.email_template.subject)
          tinyMCE.activeEditor.setContent(data.email_template.body)
        }
      })
    }else {
      $('#birs_email_subject').val('')
      tinyMCE.activeEditor.setContent('')
    }
  }

  emailModal() {
    event.preventDefault()

    var proposalIds = [];
    $("input:checked").each(function() {
      proposalIds.push(this.dataset.value);
    });
    if(typeof proposalIds[0] === "undefined")
    {
      toastr.error("Please select any checkbox!")
    }
    else {
      let id = event.currentTarget.id
      let _this = this
      $.post(`/emails/email_types?type=${id}`, function(data) {
        const selectBox = _this.templatesTarget;
        selectBox.innerHTML = '';
        const opt = document.createElement('option');
        opt.innerHTML = ''
        selectBox.appendChild(opt);
        data.email_templates.forEach((item) => {
          const opt = document.createElement('option');
          opt.innerText = item
          selectBox.appendChild(opt);
        });
        $("#email-template").modal('show')
      })
    }
  }

  sendEmails(event) {
    event.preventDefault();

    var proposalIds = [];
    $("input:checked").each(function() {
      proposalIds.push(this.dataset.value);
    });
    if(this.templatesTarget.value) {
      $('#birs_email_body').val(tinyMCE.get('birs_email_body').getContent())
      $.post(`/submitted_proposals/approve_decline_proposals?proposal_ids=${proposalIds}`,
        $("#approve_decline_proposals").serialize(), function() {
          toastr.success("Emails have been sent!")
          setTimeout(function() {
            window.location.reload();
          }, 2000)
        }
      )
      .fail(function(response) {
        let errors = response.responseJSON
        $.each(errors, function(index, error) {
          toastr.error(error)
        })
      });
    }
    else {
      toastr.error("Please select any template")
    }
  }

  tableOfContent() {
    if(this.hasTocTarget) {
      this.tocTarget.checked = true;
    }
    var proposalIds = [];
    $("input:checked").each(function() {
      proposalIds.push(this.dataset.value);
    });
    if(typeof proposalIds[1] === "undefined")
    {
      toastr.error("Please select any checkbox!")
    }
    else {
      $.post(`/submitted_proposals/table_of_content?proposals=${proposalIds}`,
        $('form#submitted-proposal').serialize(), function(data) {
          $('#proposals').text(data.proposals)
          $("#table-window").modal('show')
        }
      )
    }
  }

  booklet() {
    let ids = $('#proposals').text().slice(1)
    let table = ''
    if(this.tocTarget.checked) {
      table = "toc"
    }
    else {
      table = "ntoc"
    }
    if(table !== '') {
      document.getElementById('spinner').classList.add("active")
      $.post(`/submitted_proposals/proposals_booklet?proposal_ids=${ids}&table=${table}`,
        function() {
          document.getElementById("proposal_booklet").click();
          toastr.success('Proposals book successfully created.')
      }).fail(function() {
        toastr.error('Something went wrong.')
      })
      .always(function() {
        document.getElementById('spinner').classList.remove("active")
      })
    }
  }

  handleStatus() {
    let currentProposalId = event.currentTarget.dataset.id
    for(var i = 0; i < this.statusOptionsTargets.length; i++){
      if(currentProposalId === this.statusOptionsTargets [`${i}`].dataset.id){
        this.proposalStatusTargets [`${i}`].classList.add("hidden")
        this.statusOptionsTargets [`${i}`].classList.remove("hidden")
      }
    }
  }

  proposalStatuses() {
    let id = event.currentTarget.dataset.id
    let status = ''
    let _this = this
    for(var i = 0; i < this.statusTargets.length; i++){
      if(id === this.statusTargets [`${i}`].dataset.id){
        status = this.statusTargets [`${i}`].value
        $.post(`/submitted_proposals/${id}/update_status?status=${status}`, function() {
          toastr.success('Proposal status has been updated!')
          setTimeout(function() {
            window.location.reload();
          }, 1000)
        })
        .fail(function(res) {
          res.responseJSON.forEach((msg) => toastr.error(msg))
        });
      }
    }
  }

  storeID() {
    this.proposalId = event.currentTarget.dataset.value
  }

  selectAllProposals() {
    let getId = ''
    $("input:checkbox").each(function(){
      getId = document.getElementById(this.id)
      getId.checked = true
    });
  }

  unselectAllProposals() {
    let getId = ''
    $("input:checkbox").each(function(){
      getId = document.getElementById(this.id)
      getId.checked = false
    });
  }

  invertSelectedProposals() {
    let checkbox = ''
    $("input:checkbox").each(function(){
      checkbox = document.getElementById(this.id)
      if(checkbox.checked) {
        checkbox.checked = false
      } else {
        checkbox.checked = true
      }
    });
  }

  downloadCSV() {
    var proposalIds = [];
    $("input:checked").each(function() {
      proposalIds.push(this.dataset.value);
    });
    if(typeof proposalIds[0] === "undefined")
    {
      toastr.error("Please select any checkbox!")
    }
    else {
      let selectedProposals = proposalIds.filter((x) => typeof x !== "undefined")
      window.location = `/submitted_proposals/download_csv.csv?ids=${selectedProposals}`
    }
  }

  importReviews() {
    var proposalIds = [];
    $("input:checked").each(function() {
      proposalIds.push(this.dataset.value);
    });
    if(typeof proposalIds[0] === "undefined")
    {
      toastr.error("Please select any checkbox!")
    }
    else {
      $.post(`/submitted_proposals/import_reviews?proposals=${proposalIds}`, function(response) {
        let res = JSON.parse(response)
        if(res.type === "alert") {
          toastr.error(res.message)
        }
        else{
          toastr.success(res.message)
          setTimeout(function() {
            window.location.reload();
          }, 1000)
        }
      }).fail(function(response) {
        toastr.error(response.responseText)
      })
    }
  }

  reviewsContent() {
    if(this.hasBothReviewsTarget && this.hasReviewTocTarget) {
      this.bothReviewsTarget.checked = true;
      this.reviewTocTarget.checked = true;
    }
    var proposalIds = [];
    $("input:checked").each(function() {
      proposalIds.push(this.dataset.value);
    });
    if(typeof proposalIds[0] === "undefined")
    {
      toastr.error("Please select any checkbox!")
    }
    else {
      $("#review-window").modal('show')
    }
  }

  reviewsBooklet() {
    var proposalIds = [];
    $("input:checked").each(function() {
      proposalIds.push(this.dataset.value);
    });
    this.checkReviewType(proposalIds)
  }

  checkReviewType(proposalIds) {
    let reviewContentType = ''
    if(this.bothReviewsTarget.checked) {
      reviewContentType = "both"
    }
    else if(this.scientificReviewsTarget.checked) {
      reviewContentType = "scientific"
    }
    else {
      reviewContentType = "edi"
    }
    this.checkTableContentType(reviewContentType, proposalIds)
  }

  checkTableContentType(reviewContentType, proposalIds) {
    let table = ''
    if(this.reviewTocTarget.checked) {
      table = "toc"
    }
    else {
      table = "ntoc"
    }
    this.createReviewsBooklet(reviewContentType, proposalIds, table)
  }

  createReviewsBooklet(reviewContentType, proposalIds, table) {
    if(reviewContentType !== '') {
      document.getElementById('spinner').classList.add("active")

      $.ajax({
        url: `/submitted_proposals/reviews_booklet`,
        type: 'POST',
        data: {
          'proposals': proposalIds,
          table,
          reviewContentType
        },
        success: () => {
          document.getElementById("reviews_booklet").click();
          toastr.success('Review Booklet successfully created.')
        },
        error: () => {
          toastr.error('Something went wrong.')
        },
        complete: () => {
          document.getElementById('spinner').classList.remove("active")
        }
      })
    }
    else {
      toastr.error('Something went wrong.')
    }
  }

  removeFile(evt) {
    let dataset = evt.currentTarget.dataset
    
    $.ajax({
      url: `/reviews/${dataset.reviewId}/remove_file?attachment_id=${dataset.attachmentId}`,
      type: 'DELETE',
      data: {
        'attachment_id': dataset.attachmentId
      },
      success: () => {
        $(`#review-file${dataset.attachmentId}`).remove()
        toastr.success('Comment has successfully been removed.')
      },
      error: () => {
        toastr.error('Something went wrong.')
      }
    })
  }

  addFile(evt) {
    let dataset = evt.currentTarget.dataset
    if(evt.target.files) {
      var data = new FormData()
      var f = evt.target.files[0]
      var ext = f.name.split('.').pop();
      this.sendRequest(ext, data, f, dataset)
    }
  }

  reviewsExcelBooklet() {
    var proposalIds = [];
    $("input:checked").each(function() {
      proposalIds.push(this.dataset.value);
    });
    if(typeof proposalIds[0] === "undefined")
    {
      toastr.error("Please select any checkbox!")
    }
    else {
      window.location = `/submitted_proposals/reviews_excel_booklet.xlsx?proposals=${proposalIds}`
    }
  }

  sendRequest(ext, data, f, dataset) {
    if( ext === "pdf" || ext === "txt" || ext === "text") {
      data.append("file", f)
      var url = `/reviews/${dataset.reviewId}/add_file`
      Rails.ajax({
        url,
        type: "POST",
        data,
        success: () => {
          location.reload(true)
          toastr.success('File is attached successfully.')
        },
        error: (response) => {
          toastr.error(response.errors)
        }
      })
    }
    else {
      toastr.error('Only .pdf and .txt files are allowed.')
    }
  }
}
